test/unit-tests/cpptools.test.ts

```diff
@@ -61,20 +61,58 @@ suite('CppTools tests', () => {
       }]
     };
 
-    provider.updateConfigurationData({cache, codeModel, activeTarget: 'target1'});
+    provider.updateConfigurationData({cache, codeModel, activeTarget: 'target1', folder: here});
     let configurations = await provider.provideConfigurations([uri]);
     expect(configurations.length).to.eq(1);
     expect(configurations[0].configuration.defines).to.contain('FLAG1');
 
-    provider.updateConfigurationData({cache, codeModel, activeTarget: 'target2'});
+    provider.updateConfigurationData({cache, codeModel, activeTarget: 'target2', folder: here});
     configurations = await provider.provideConfigurations([uri]);
     expect(configurations.length).to.eq(1);
     expect(configurations[0].configuration.defines).to.contain('FLAG2');
 
-    provider.updateConfigurationData({cache, codeModel, activeTarget: 'all'});
+    provider.updateConfigurationData({cache, codeModel, activeTarget: 'all', folder: here});
     configurations = await provider.provideConfigurations([uri]);
     expect(configurations.length).to.eq(1);
     expect(configurations[0].configuration.defines.some(def => def === 'FLAG1' || def === 'FLAG2')).to.be.true;
     expect(configurations[0].configuration.defines).to.not.have.all.members(['FLAG1', 'FLAG2']);
+
+    // Verify the per-folder browsePath.
+    let canProvideBrowseConfigPerFolder: boolean = await provider.canProvideBrowseConfigurationsPerFolder();
+    expect(canProvideBrowseConfigPerFolder).to.eq(true);
+    let browseConfig = await provider.provideFolderBrowseConfiguration(vscode.Uri.file(path.join(here, __filename)));
+    expect(browseConfig.browsePath.length).to.eq(1);
+    expect(browseConfig.browsePath[0]).to.eq(here);
+
+    // Verify the browsePath with a different folder.
+    let smokeFolder = path.join(here, '../smoke');
+    const sourceFile2 = path.join(smokeFolder, 'main.cpp');
+    const uri2 = vscode.Uri.file(sourceFile2);
+    const codeModel2: codemodel_api.CodeModelContent = {
+      configurations: [{
+        projects: [{
+          name: 'cpptools-test2',
+          sourceDirectory: smokeFolder,
+          targets: [
+            {
+              name: 'target1',
+              type: 'EXECUTABLE',
+              fileGroups: [{
+                sources: [sourceFile2],
+                isGenerated: false,
+                compileFlags: '-DFLAG1',
+                language: 'CXX'
+              }]
+            }]
+        }]
+      }]
+    };
+    provider.updateConfigurationData({cache, codeModel: codeModel2, activeTarget: 'target1', folder: smokeFolder});
```

Comment:
I think it would be better to pass in all of the code models up front before doing any validation.  This would help validate that the different folder states do not pollute the API responses.

Reply:
Ok